pandas/core/algorithms.py

```diff
@@ -619,13 +619,19 @@ def factorize(values, sort=False, order=None, na_sentinel=-1, size_hint=None):
 
     if sort and len(uniques) > 0:
         from pandas.core.sorting import safe_sort
-        try:
-            order = uniques.argsort()
-            order2 = order.argsort()
-            labels = take_1d(order2, labels, fill_value=na_sentinel)
-            uniques = uniques.take(order)
-        except TypeError:
-            # Mixed types, where uniques.argsort fails.
+        if na_sentinel == -1:
+            # GH-25409 take_1d only works for na_sentinels of -1
+            try:
+                order = uniques.argsort()
+                order2 = order.argsort()
+                labels = take_1d(order2, labels, fill_value=na_sentinel)
+                uniques = uniques.take(order)
+            except TypeError:
+                # Mixed types, where uniques.argsort fails.
+                uniques, labels = safe_sort(uniques, labels,
```

Comment:
I don't like the way this is written because it repeats the call to safe_sort. if you can avoid that (easy enough, just use a pass in the except), *then* call safe_sort if needed (assign to uniques, lables as NOne, None initially), then this would be better

Reply:
@jorisvandenbossche do you find this clearer? Personally, I find these dummy variables just for flow control a bit confusing (you have to verify that things are eventually set somewhere). With the way things are written, it's clear that each of the three branches has a `uniques` and `labels`, and it's clear how things are achieved.

If we're really concerned about repeating the call to safe_sort, you can make a closer with the values

```
get_uniques_labels = lambda: safe_sort(uniques, labels, na_sentinel, assume_unique=True)`
```

and call the lambda, but again, that's just unnecessary indirection IMO.

In this case I think just do whatever.

Reply:
I agree with Tom that this is not necessarily clearer as what it is now.